sentinel-adapter/sentinel-zuul2-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/zuul2/filters/inbound/SentinelZuulInboundFilter.java

```diff
@@ -0,0 +1,151 @@
+import rx.Observable;
+import rx.schedulers.Schedulers;
+
+import static com.alibaba.csp.sentinel.adapter.gateway.common.SentinelGatewayConstants.*;
+
+/**
+ * Zuul2 inboundFilter for Sentinel.
+ *
+ * @author wavesZh
+ */
+public class SentinelZuulInboundFilter extends HttpInboundFilter {
+
+    private static final String DEFAULT_BLOCK_ENDPOINT_NAME = SentinelZuulEndpoint.class.getCanonicalName();
+
+    private final int order;
+
+    private final String blockedEndpointName;
+    /**
+     * if executor is null, flow control action will do on I/O thread
+     */
+    private final Executor executor;
+
+    private final GatewayParamParser<HttpRequestMessage> paramParser = new GatewayParamParser<>(
+            new HttpRequestMessageItemParser());
+
+    public SentinelZuulInboundFilter(int order) {
+        this(order, null);
+    }
+
+	public SentinelZuulInboundFilter(int order, Executor executor) {
+        this(order, DEFAULT_BLOCK_ENDPOINT_NAME, executor);
+	}
+
+    public SentinelZuulInboundFilter(int order, String blockedEndpointName, Executor executor) {
+        this.order = order;
+		this.blockedEndpointName = blockedEndpointName;
+		this.executor = executor;
+	}
+
+    @Override
+    public int filterOrder() {
+        return order;
+    }
+
+    @Override
+    public Observable<HttpRequestMessage> applyAsync(HttpRequestMessage request) {
+        if (executor != null) {
+            return Observable.just(request).subscribeOn(Schedulers.from(executor)).flatMap(this::apply);
+        } else {
+            return Observable.just(request).flatMap(this::apply);
+        }
+    }
+
+    private Observable<HttpRequestMessage> apply(HttpRequestMessage request) {
+        SessionContext context = request.getContext();
+        String origin = parseOrigin(request);
+        Deque<AsyncEntry> asyncEntries = new ArrayDeque<>();
+        String routeId = (String) context.get(ZuulConstant.PROXY_ID_KEY);
+        String fallBackRoute = routeId;
+        try {
+            if (StringUtil.isNotBlank(routeId)) {
+                ContextUtil.enter(GATEWAY_CONTEXT_ROUTE_PREFIX + routeId, origin);
+                doSentinelEntry(routeId, RESOURCE_MODE_ROUTE_ID, request, asyncEntries);
+            }
+            Set<String> matchingApis = pickMatchingApiDefinitions(request);
+            if (!matchingApis.isEmpty() && ContextUtil.getContext() == null) {
+                ContextUtil.enter(ZuulConstant.ZUUL_DEFAULT_CONTEXT, origin);
+            }
+            for (String apiName : matchingApis) {
+                fallBackRoute = apiName;
+                doSentinelEntry(apiName, RESOURCE_MODE_CUSTOM_API_NAME, request, asyncEntries);
+            }
+            return Observable.just(request);
+        } catch (BlockException t) {
+			context.put(ZuulConstant.ZUUL_CTX_SENTINEL_BLOCKED_FLAG, Boolean.TRUE);
+			context.put(ZuulConstant.ZUUL_CTX_SENTINEL_FALLBACK_ROUTE, fallBackRoute);
+            context.setEndpoint(blockedEndpointName);
```

Comment:
It's better to call "setErrorEndpoint" and "setShouldSendErrorResponse" here.

Reply:
If you set "setShouldSendErrorResponse" with true, the rest of inbound filter will not be invoked. I want that user can control the invocation of inbound filter by `ZuulConstant.ZUUL_CTX_SENTINEL_BLOCKED_FLAG`.  It can be more flexible.

Reply:
1.Generally when a resouce is blocked,the invocation will be aborted.I think it would be confusing if the invocation continues.It seems better to make this feature enabled by configuration and disabled by default？
2.By the way I finally work out the context switching(not tested although) , may you have a look at my [code](https://github.com/locuslin/Sentinel/blob/7e013fa83c49794e80b030adbe20b5b31e5f65cf/sentinel-adapter/sentinel-zuul2-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/zuul2/filters/SentinelZuulInboundFilter.java#L104)?

Reply:
1. Your suggestion is well, i will optimize it.
2. I see a little your code,  what is meaning about [the code](https://github.com/locuslin/Sentinel/blob/7e013fa83c49794e80b030adbe20b5b31e5f65cf/sentinel-adapter/sentinel-zuul2-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/gateway/zuul2/filters/SentinelZuulOutboundFilter.java#L49). asynchronous action should be done in asynchronous context to keep a correct invocation chain.

Reply:
Just refer to the `SentinelReactorSubscriber`,exactly [hookOnSubscribe](https://github.com/alibaba/Sentinel/blob/73188b46b78223c5d075b2cb9444a87efec192f0/sentinel-adapter/sentinel-reactor-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/reactor/SentinelReactorSubscriber.java#L111) and [hookOnNext](https://github.com/alibaba/Sentinel/blob/73188b46b78223c5d075b2cb9444a87efec192f0/sentinel-adapter/sentinel-reactor-adapter/src/main/java/com/alibaba/csp/sentinel/adapter/reactor/SentinelReactorSubscriber.java#L117).It's not necessary to do like this,but just in case and to keep the same style.